nc/source/physics/dynamics/Solver.cpp

```diff
@@ -0,0 +1,227 @@
+                               const ConstraintMatrix& jNormal,
+                               const ConstraintMatrix& jTangent,
+                               const ConstraintMatrix& jBitangent)
+    {
+        auto dot0 = XMVector3Dot(jNormal.vA(), v.vA());
+        auto dot1 = XMVector3Dot(jNormal.wA(), v.wA());
+        auto dot2 = XMVector3Dot(jNormal.vB(), v.vB());
+        auto dot3 = XMVector3Dot(jNormal.wB(), v.wB());
+        auto result = dot0 + dot1 + dot2 + dot3;
+
+        dot0 = XMVector3Dot(jTangent.vA(), v.vA());
+        dot1 = XMVector3Dot(jTangent.wA(), v.wA());
+        dot2 = XMVector3Dot(jTangent.vB(), v.vB());
+        dot3 = XMVector3Dot(jTangent.wB(), v.wB());
+        result = XMVectorMergeXY(result, dot0 + dot1 + dot2 + dot3);
+
+        dot0 = XMVector3Dot(jBitangent.vA(), v.vA());
+        dot1 = XMVector3Dot(jBitangent.wA(), v.wA());
+        dot2 = XMVector3Dot(jBitangent.vB(), v.vB());
+        dot3 = XMVector3Dot(jBitangent.wB(), v.wB());
+        result = XMVectorPermute<XM_PERMUTE_0X, XM_PERMUTE_0Y, XM_PERMUTE_1X, XM_PERMUTE_1Y>(result, dot0 + dot1 + dot2 + dot3);
+        return result;
+    }
+
+    ConstraintMatrix ComputeDeltas(const ContactConstraint& constraint, float lNormal, float lTangent, float lBitangent)
+    {
+        const auto &jN = constraint.jNormal, &jT = constraint.jTangent, &jB = constraint.jBitangent;
+        const auto &iA = constraint.invInertiaA, &iB = constraint.invInertiaB;
+        float mA = constraint.invMassA, mB = constraint.invMassB;
+
+        auto vA = XMVectorScale(jN.vA(), mA * lNormal) +
+                  XMVectorScale(jT.vA(), mA * lTangent) +
+                  XMVectorScale(jB.vA(), mA * lBitangent);
+
+        auto wA = XMVectorScale(XMVector3Transform(jN.wA(), iA), lNormal) +
+                  XMVectorScale(XMVector3Transform(jT.wA(), iA), lTangent) +
+                  XMVectorScale(XMVector3Transform(jB.wA(), iA), lBitangent);
+
+        auto vB = XMVectorScale(jN.vB(), mB * lNormal) +
+                  XMVectorScale(jT.vB(), mB * lTangent) +
+                  XMVectorScale(jB.vB(), mB * lBitangent);
+        
+        auto wB = XMVectorScale(XMVector3Transform(jN.wB(), iB), lNormal) +
+                  XMVectorScale(XMVector3Transform(jT.wB(), iB), lTangent) +
+                  XMVectorScale(XMVector3Transform(jB.wB(), iB), lBitangent);
+
+        return ConstraintMatrix{vA, wA, vB, wB};
+    }
+
+    float ClampLambda(float newLambda, float* totalLambda)
+    {
+        float temp = *totalLambda;
+        *totalLambda = math::Max(0.0f, temp + newLambda);
+        return *totalLambda - temp;
+    }
+
+    float ClampMu(float newMu, float extent, float* totalMu)
+    {
+        float temp = *totalMu;
+        *totalMu = math::Clamp(temp + newMu, -1.0f * extent, extent);
+        return *totalMu - temp;
+    }
+
+    BasicContactConstraint CreateBasicContactConstraint(Entity a, Entity b, Transform* transformA, Transform* transformB, const Manifold& manifold)
```

Comment:
I'm not a big fan of methods creating Constraints being in a class called Solver when we have a Constraint class.
Maybe make a ConstraintFactory that creates these different constraints?